torch/csrc/jit/codegen/cuda/runtime/welford.cu

```diff
@@ -0,0 +1,423 @@
+template <
+    bool X_REDUCE,
+    bool Y_REDUCE,
+    bool Z_REDUCE,
+    typename T,
+    typename _dim3ti,
+    typename _dim3bd>
+__inline__ __device__ void blockWelford(
+    T& out_M2,
+    T& out_avg,
+    int& out_N,
+    const T& in_M2,
+    const T& in_avg,
+    const int N,
+    const _dim3ti& thread_idx,
+    const _dim3bd& block_dim,
+    T* shared_mem_M2,
+    T* shared_mem_avg,
+    bool read_write_pred,
+    T init_val) {
+  unsigned int reduction_size = (X_REDUCE ? block_dim.x : 1) *
+      (Y_REDUCE ? block_dim.y : 1) * (Z_REDUCE ? block_dim.z : 1);
+  // If this thread will output a final result
+  bool should_write = true;
+  if (X_REDUCE)
+    should_write = should_write && thread_idx.x == 0;
+  if (Y_REDUCE)
+    should_write = should_write && thread_idx.y == 0;
+  if (Z_REDUCE)
+    should_write = should_write && thread_idx.z == 0;
+  unsigned int reduction_stride;
+  unsigned int reduction_tid;
+  unsigned int linear_tid;
+  if (X_REDUCE && !Y_REDUCE && Z_REDUCE) {
+    // Transpose Z and Y in the shared memory so Z and X dims are contiguous in
+    // smem
+    reduction_stride = 1;
+    linear_tid = threadIdx.y * blockDim.z * blockDim.x +
+        threadIdx.z * blockDim.x + threadIdx.x;
+    reduction_tid = threadIdx.z * blockDim.x + threadIdx.x;
+  } else {
+    // Normal reduction in order
+    reduction_stride =
+        (X_REDUCE ? 1
+                  : (Y_REDUCE ? block_dim.x
+                              : (Z_REDUCE ? block_dim.x * block_dim.y : 0)));
+    linear_tid = thread_idx.z * block_dim.y * block_dim.x +
+        thread_idx.y * block_dim.x + thread_idx.x;
+    reduction_tid = (Z_REDUCE ? thread_idx.z : 0) *
+            (Y_REDUCE ? block_dim.y : 1) * (X_REDUCE ? block_dim.x : 1) +
+        (Y_REDUCE ? thread_idx.y : 0) * (X_REDUCE ? block_dim.x : 1) +
+        (X_REDUCE ? thread_idx.x : 0);
+  }
+  // assert( reduction_stride != 0 );
+  if (read_write_pred) {
+    shared_mem_M2[linear_tid] = in_M2;
+    shared_mem_avg[linear_tid] = in_avg;
+  } else {
+    shared_mem_M2[linear_tid] = init_val;
+    shared_mem_avg[linear_tid] = init_val;
+  }
+  __syncthreads();
+  // Reduce down to nearest power of 2:
+  int np2 = 1 << (31 - __clz(reduction_size));
+  int lower = N;
+  int upper = 0;
+  int border = reduction_size;
+  if (reduction_tid < np2) {
+    if (reduction_tid + np2 < reduction_size) {
+      int tmp_ = getCount(lower, upper, border, reduction_tid);
+      welfordCombine(
+          shared_mem_M2[linear_tid],
+          shared_mem_avg[linear_tid],
+          tmp_,
+          shared_mem_M2[linear_tid + np2 * reduction_stride],
+          shared_mem_avg[linear_tid + np2 * reduction_stride],
+          getCount(lower, upper, border, reduction_tid + np2));
+    }
+    foldCount(lower, upper, border, np2);
+  }
+  __syncthreads();
+  for (int factor = np2 / 2; factor > 0; factor >>= 1) {
+    if (reduction_tid < factor) {
+      int tmp_ = getCount(lower, upper, border, reduction_tid);
```

Comment:
Looks like we are indeed updating `tmp_` in `welfordCombine`.. I think we should take this outside of the for loop.

Reply:
Thanks for taking a deeper look into this. I agree that this part of implementation is not as intuitive as I hoped. The main purpose of this effort was not to fill the counts in `shared_mem` because I just felt it be nice not to fill the mem with all identical numbers.

Reply:
In block reduce the actual count was managed by `foldCount` and `tmp_` was just there to supply the number as an argument to `welfordCombine`. Using `tmp_` because `welfordCombine` does pass-by-reference and we don't need the modified value in this case. `foldCount` would calculate the count value for all the positions in each iterartion.

Reply:
Oops, sorry I missed that part, got confused between the grid/block version :stuck_out_tongue: 
Nice trick on `foldCount`.